src/Tools/ExternalAccess/Razor/Remote/RazorHostOutputHandler.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Linq;
+using System.Runtime.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ExternalAccess.RazorCompiler;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Remote
+{
+    [DataContract]
+    internal class HostOutputRequest
+    {
+        [DataMember(Name = "textDocument")]
+        internal required TextDocumentIdentifier TextDocument { get; set; }
+
+        [DataMember(Name = "generatorName")]
+        internal required string GeneratorName { get; set; }
+
+        [DataMember(Name = "requestedOutput")]
+        internal required string RequestedOutput { get; set; }
+    }
+
+    [DataContract]
+    internal class HostOutputResponse
+    {
+        [DataMember(Name = "outputs")]
+        internal string? Output { get; set; }
+    }
+
+    [Method(MethodName)]
+    [ExportMetadata("Extensions", new string[] { "cshtml", "razor" })]
+    [ExportCSharpVisualBasicStatelessLspServiceAttribute(typeof(RazorHostOutputHandler)), Shared]
+    internal class RazorHostOutputHandler : ILspServiceRequestHandler<HostOutputRequest, HostOutputResponse>
+    {
+        internal const string MethodName = "razor/hostOutput";
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public RazorHostOutputHandler()
+        {
+        }
+
+        public bool MutatesSolutionState { get; } = false;
+
+        public bool RequiresLSPSolution { get; } = true;
+
+        public async Task<HostOutputResponse> HandleRequestAsync(HostOutputRequest request, RequestContext context, CancellationToken cancellationToken)
+        {
+            string? output = null;
+
+            var project = context.Solution?.GetAdditionalDocument(request.TextDocument)?.Project;
+            if (project is not null)
+            {
+                output = await GetHostOutputAsync(project, request.GeneratorName, request.RequestedOutput, cancellationToken).ConfigureAwait(false);
+            }
+            return new HostOutputResponse() { Output = output };
+        }
+
+        public static async Task<string?> GetHostOutputAsync(Project project, string generatorName, string requestedOutput, CancellationToken cancellationToken)
```

Comment:
I rather wish this was the method we were introducing on the CompilationTrackers directly, since that way we're not potentially exposing a richer object that has other subtle behaviors.

Reply:
(see my other comment about your PROTOTYPE comment.)

Reply:
@jasonmalinowski I took a look at doing this, but the issue is that `GetHostOutput` is internal to the compiler, we expose it via IVT to razor via External Access.

If we wanted to return the host outputs directly, we'd need to make workspaces aware of it somehow, either by making the API public or use of IVT, which as I understand it, is generally considered verboten.

Reply:
@chsienki Ick. Yes, workspace to IVT is verboten. I'm trying to remember if we had a plan in place for dealing with something like this. Or the other answer is we go with this as is, and once we've proved it out a bit further we can then make this "proper", but I hugely worry about the debt being piled on for Razor, generally.

Reply:
@jasonmalinowski 

>  but I hugely worry about the debt being piled on for Razor, generally.

Quite. One of my hopes with some of this change is that we can actually start